service/history/queues/queue_base.go

```diff
@@ -339,61 +339,75 @@ func (p *queueBase) checkpoint() {
 		}
 	}()
 
-	totalSlices := 0
 	readerScopes := make(map[int32][]Scope)
-	newInclusiveLowWatermark := tasks.MaximumKey
+	newInclusiveLowWatermark := p.nonReadableScope.Range.InclusiveMin
 	for id, reader := range p.readerGroup.Readers() {
 		reader.ShrinkSlices()
 		scopes := reader.Scopes()
 
-		totalSlices += len(scopes)
 		readerScopes[id] = scopes
 		for _, scope := range scopes {
 			newInclusiveLowWatermark = tasks.MinKey(newInclusiveLowWatermark, scope.Range.InclusiveMin)
 		}
 	}
 	p.metricsHandler.Histogram(QueueReaderCountHistogram, metrics.Dimensionless).Record(int64(len(readerScopes)))
-	p.metricsHandler.Histogram(QueueSliceCountHistogram, metrics.Dimensionless).Record(int64(totalSlices))
+	p.metricsHandler.Histogram(QueueSliceCountHistogram, metrics.Dimensionless).Record(int64(p.monitor.GetTotalSliceCount()))
+	p.metricsHandler.Histogram(PendingTasksCounter, metrics.Dimensionless).Record(int64(p.monitor.GetTotalPendingTaskCount()))
 
 	// NOTE: Must range complete task first.
 	// Otherwise, if state is updated first, later deletion fails and shard get reloaded
 	// some tasks will never be deleted.
-	if newInclusiveLowWatermark != tasks.MaximumKey && newInclusiveLowWatermark.CompareTo(p.inclusiveLowWatermark) > 0 {
-		p.metricsHandler.Counter(TaskBatchCompleteCounter).Record(1)
-
-		persistenceMinTaskKey := p.inclusiveLowWatermark
-		persistenceMaxTaskKey := newInclusiveLowWatermark
-		if p.category.Type() == tasks.CategoryTypeScheduled {
-			persistenceMinTaskKey = tasks.NewKey(p.inclusiveLowWatermark.FireTime, 0)
-			persistenceMaxTaskKey = tasks.NewKey(newInclusiveLowWatermark.FireTime, 0)
-		}
-
-		ctx, cancel := newQueueIOContext()
-		defer cancel()
-
-		if err = p.shard.GetExecutionManager().RangeCompleteHistoryTasks(ctx, &persistence.RangeCompleteHistoryTasksRequest{
-			ShardID:             p.shard.GetShardID(),
-			TaskCategory:        p.category,
-			InclusiveMinTaskKey: persistenceMinTaskKey,
-			ExclusiveMaxTaskKey: persistenceMaxTaskKey,
-		}); err != nil {
-			p.logger.Error("Error range completing queue task", tag.Error(err))
+	if newInclusiveLowWatermark.CompareTo(p.inclusiveLowWatermark) > 0 {
+		err = p.rangeCompleteTasks(p.inclusiveLowWatermark, newInclusiveLowWatermark)
+		if err != nil {
 			return
 		}
 
 		p.inclusiveLowWatermark = newInclusiveLowWatermark
 	}
 
+	err = p.updateQueueState(readerScopes)
+}
+
+func (p *queueBase) rangeCompleteTasks(
+	oldInclusiveMinTaskKey tasks.Key,
+	newInclusiveMinTaskKey tasks.Key,
+) error {
+	p.metricsHandler.Counter(TaskBatchCompleteCounter).Record(1)
```

Comment:
why need this one? Can we use persistence_request {operation=RangeCompleteHistoryTasks}?

Reply:
This is emitted by the existing non-multicursor implementation so I think I'd better keep it in case any user is relying on the metrics. 

I double checked the existing implementation, this metric is emitted before checking whether or not the deletion level changed or not. So if a shard has not task this metric will still be emitted but there'll be not persistence request metric. I've also moved the emission here out of this function.